### PR TITLE
Fix bincount to use acc scalar for the bounds

### PR DESCRIPTION
**Pull request by t-vi**
The bounds could overflow when the number of bins is larger than the type can use, e.g. when uint8 inputs want 256 bins.

Thank you, Yang Xiaobo, for reporting a reproducing example in the forums.

### CODE REVIEW

**Comment by mruberry** [comment]
Cool! Looks like a smart fix to me, @ngimel?

File: aten/src/ATen/native/cuda/SummaryOps.cu [hunk 1]
```diff
@@ -17,6 +18,8 @@
 #include <ATen/ops/zeros_native.h>
 #endif
 
+#include <iostream>
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/SummaryOps.cu:21]: stray?

File: aten/src/ATen/native/cuda/SummaryOps.cu [hunk 2]
```diff
@@ -405,8 +438,11 @@ Tensor _bincount_cuda(
   globalContext().alertNotDeterministic("_bincount_cuda");
   return AT_DISPATCH_INTEGRAL_TYPES(self.scalar_type(), "bincount_cuda", [&] {
     const auto scalar = weights.scalar_type();
-    if (scalar == ScalarType::Undefined || scalar == ScalarType::Float)
+    if (scalar == ScalarType::Undefined || scalar == ScalarType::Float) {
+      std::cerr << "float!\n";
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/SummaryOps.cu:442]: stray print? (and another one below)

File: aten/src/ATen/native/cuda/SummaryOps.cu [hunk 3]
```diff
@@ -31,16 +34,22 @@ namespace cuda {
...
-  __device__ static IndexType getBin(input_t bVal, input_t minvalue, input_t maxvalue, int64_t nbins) {
-    IndexType bin = (int)((bVal - minvalue) * nbins / (maxvalue - minvalue));
-    // (only applicable for histc)
-    // while each bin is inclusive at the lower end and exclusive at the higher, i.e. [start, end)
-    // the last bin is inclusive at both, i.e. [start, end], in order to include maxvalue if exists
-    // therefore when bin == nbins, adjust bin to the last bin
-    if (bin == nbins) bin -= 1;
-    return bin;
-  }
+template <typename input_t, typename bounds_t, typename IndexType>
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/SummaryOps.cu:37]: do you need to string through another template argument, or could you have just declared `acc_type<input_t, true> minvalue`?

**Review by ngimel — COMMENTED** [review]
This looks cool, but please remove stray prints

File: aten/src/ATen/native/cuda/SummaryOps.cu (hunk 2 above)
Comment (t-vi) [reply aten/src/ATen/native/cuda/SummaryOps.cu:442]: Ohoh, caught me. :) Thank you!

File: aten/src/ATen/native/cuda/SummaryOps.cu (hunk 3 above)
Comment (t-vi) [reply aten/src/ATen/native/cuda/SummaryOps.cu:37]: To be honest, neither thing is aesthetically satisfying, to the point where I couldn't tell the difference. But in lieu of my own taste, I'll gladly go with yours. Thank you!

Comment (ngimel) [reply aten/src/ATen/native/cuda/SummaryOps.cu]: I agree, neither is aesthetically pleasing, but I have a reflexive reaction to extra template argument "Oh, you are instantiating new templates" even though you aren't, and some code digging is required to make sure that no new instantiations happen.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**